File: leetcode3/변지협/v2/948. Bag of Tokens.py

'''
1. 아이디어 :
점수를 잃을땐 최대 power를 얻고, 점수를 얻을 땐 최소 pwoer를 소모한다.

2. 시간복잡도 :
o(nlogn)

3. 자료구조/알고리즘 :
투포인터
'''


class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        tokens.sort()
        n = len(tokens)
        
        a = 0
        b = n - 1
        score = 0
        ans = 0

        while True:
            print('a,b,power,score:', a,b,power)
            if a > b:
                break
            
            if power >= tokens[a]:
                power -= tokens[a]
                score +=1
                ans = max([score, ans])
                a += 1
            else:
                if score == 0:
                    break
                power += tokens[b]
                score -= 1
                b -= 1
        
        return ans

File: leetcode3/황은지/948. Bag of Tokens.js
/**
 * @param {number[]} tokens
 * @param {number} power
 * @return {number}
 */
var bagOfTokensScore = function (tokens, power) {
  const dp = Array(10000);
  dp[power] = 0;
  const queue = [[power, 0]];
  let head = 0;
  let maxScore = 0;

  while (queue.length - head > 0) {
    const [curP, curS] = queue[head++];
    maxScore = Math.max(maxScore, curS);

    for (const token of tokens) {
      if (curP >= token) queue.push([curP - token, curS + 1]);
      if (curS >= 1) queue.push([curP + token, curS - 1]);
    }
  }
};
